### PR TITLE
Fhmap 451 add new forms api

### PR DESCRIPTION
**Pull request by nialldonnellyfh**
The Appforms APIs have all been refactored to use the new command format for fhc.

Additionally, the API includes all of the commands necessary for performing LCM management actions.

@cianclarke if you get a chance, can you take a look.

### CODE REVIEW

**Comment by fheng-build-servers** [comment]
@cianclarke @david-martin please review this PR.

File: lib/cmd/fh3/appforms/environments/submissions/create.js [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+      var submissionJSON;
+      try{
+        submissionJSON = JSON.parse(submissionFileContent);
+      } catch(e){
+        return cb("Invalid Submission JSON Object");
+      }
+
+      //If the user supplies a project id, then assign it to the submission JSON data
+      if(params.projectid){
+        submissionJSON.appId = params.projectid
```
Comment (cianclarke) [inline lib/cmd/fh3/appforms/environments/submissions/create.js]: potential for submissionJSON to be `undefined`?

Comment (cianclarke) [inline lib/cmd/fh3/appforms/environments/submissions/create.js]: Nvm - submissionFileContent  falls back to "", and JSON.parse fails on `JSON.parse('')`

File: lib/cmd/fh3/appforms/environments/submissions/exportcsv.js [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+  'customCmd': function(params, cb){
+    //Custom commad to make a file request.
+    params.url = this.url(params);
+    params.method = this.method;
+
+    if(params.output.indexOf(".zip") === -1){
+      return cb("Expected The Output File To Have A .zip Extension");
+    }
+
+    fs.exists(params.output, function(exists){
```
Comment (cianclarke) [inline lib/cmd/fh3/appforms/environments/submissions/exportcsv.js]: Isn't fs.exists due for deprecation?
https://nodejs.org/api/fs.html#fs_fs_exists_path_callback

Hasn't yet been marked as such, but not sure if it'd throw some sort of warnings anywhere? Feel free to convert to `fs.stat` or similar if you want, but no panic!

File: lib/cmd/fh3/appforms/environments/submissions/exportpdf.js [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+  'customCmd': function(params, cb){
+    //Custom commad to make a file request.
+    params.url = this.url(params);
+    params.method = this.method;
+
+    if(params.output.indexOf(".pdf") === -1){
+      return cb("Expected The Output File To Have A .pdf Extension");
+    }
+
+    fs.exists(params.output, function(exists){
```
Comment (cianclarke) [inline lib/cmd/fh3/appforms/environments/submissions/exportpdf.js]: As above note on `fs.exists`

File: lib/cmd/fh3/appforms/environments/submissions/getfile.js [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+  },
+  'url' : function(params){
+    return "/api/v2/mbaas/" + params.environment + "/appforms/submissions/" + params.id + "/files/" + params.fileid;
+  },
+  'customCmd': function(params, cb){
+    //Custom commad to make a file request, in this case, the returned file is a PDF.
+    params.url = this.url(params);
+    params.method = this.method;
+
+    fs.exists(params.output, function(exists){
```
Comment (cianclarke) [inline lib/cmd/fh3/appforms/environments/submissions/getfile.js]: Another `fs.exists`

File: lib/cmd/fh3/appforms/environments/submissions/update.js [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+  },
+  'url' : function(params){
+    return "/api/v2/mbaas/" + params.environment + "/appforms/submissions/" + params.id;
+  },
+  'preCmd' : function(params, cb){
+    var submissionDataPath = params.submissiondata;
+
+    //Reading The File Passed For Form Content
+    fs.readFile(submissionDataPath, function(err, submissionFileContent){
+      submissionFileContent = submissionFileContent || "";
```
Comment (cianclarke) [inline lib/cmd/fh3/appforms/environments/submissions/update.js]: Maybe check for err here? 
I know submissionFileContent will cause the try/catch to fail, but a more meaningful error will be populated here than below

File: lib/cmd/fh3/appforms/environments/submissions/exportcsv.js (hunk 2 above)
Comment (nialldonnellyfh) [inline lib/cmd/fh3/appforms/environments/submissions/exportcsv.js]: Good catch, will replace these.

File: lib/cmd/fh3/appforms/environments/submissions/create.js [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+    'projectid': "Project Guid To Associate The Submission With"
+  },
+  'url' : function(params){
+    return "/api/v2/mbaas/" + params.environment + "/appforms/submissions";
+  },
+  'preCmd' : function(params, cb){
+    var submissionDataPath = params.submissiondata;
+
+    //Reading The File Passed For Form Content
+    fs.readFile(submissionDataPath, function(err, submissionFileContent){
```
Comment (cianclarke) [inline lib/cmd/fh3/appforms/environments/submissions/create.js]: as below, maybe consider checking for err as well as relying on try/catch handler?

File: lib/cmd/fh3/appforms/forms/create.js [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+  'describe' : {
+    'formfile' : 'A Path To A File Containing A Valid JSON Definition Of A Form'
+  },
+  'url' : '/api/v2/appforms/forms',
+  'method' : 'post',
+  'preCmd' : function(params, cb){
+    var formFilePath = params.formfile;
+    //Reading The File Passed For Form Content
+    fs.readFile(formFilePath, function(err, formFileContent){
+      formFileContent = formFileContent || "";
```
Comment (cianclarke) [inline lib/cmd/fh3/appforms/forms/create.js]: as before, check for err maybe

File: lib/utils/request.js [hunk 8]
```diff
@@ -193,6 +192,72 @@ function MULTIGET (fhurls, where, cb) {
...
 
+/**
+ * Utility function to download files to a file path from a url
+ * @param params
+ * @param cb
+ */
+function downloadFile(params, cb){
+  var url = params.url;
+  var filePath = params.output;
+  var fileStream = fs.createWriteStream(filePath);
```
Comment (cianclarke) [inline lib/utils/request.js]: Should fileStream also have an error listner? What happens if the specified filePath doesn't exist, or permissions are lacking?

File: npm-shrinkwrap.json [hunk 9]
```diff
@@ -1,11 +1,11 @@
 {
   "name": "fh-fhc",
-  "version": "1.0.2-BUILD-NUMBER",
+  "version": "2.0.0-BUILD-NUMBER",
```
Comment (cianclarke) [inline npm-shrinkwrap.json]: Are we breaking the previous forms FHC APIs with this change? Should this definitely be a major bump? 
I'm also thinking yes, but just wanted to raise the question!

**Comment by cianclarke** [comment]
` 1,725 additions and 1,327 deletions.` - great stuff, with added functionality barely 400 lines extra. The new approach to commands seems to help DRY things up, this is probably the biggest incumbent set of commands yet. 
Reviewed, few tiny things and that's it. nicely done

File: npm-shrinkwrap.json (hunk 9 above)
Comment (nialldonnellyfh) [inline npm-shrinkwrap.json]: Yep, all the API urls have changed.

File: lib/utils/request.js (hunk 8 above)
Comment (nialldonnellyfh) [inline lib/utils/request.js]: The error should propagate up the pipe. Otherwise, we get the risk of doing a double callback.

**Pull request closed without merging**